include/kangaru.hpp

```diff
@@ -97,43 +106,50 @@ struct Container : std::enable_shared_from_this<Container> {
 	template<int S, typename T> using parent_element = typename std::tuple_element<S, parent_types<T>>::type;
 	template<int S, typename Tuple> using tuple_element = typename std::tuple_element<S, Tuple>::type;
 	using holder_ptr = std::unique_ptr<detail::Holder>;
-
+	using holder_cont = std::unordered_map<detail::type_id_fn, holder_ptr>;
 public:
+	Container() = default;
+	Container(const Container &) = default;
+	Container(Container &&) = default;
+	Container& operator =(const Container &) = default;
+	Container& operator =(Container &&) = default;
+	virtual ~Container() = default;
+
 	template<typename T>
 	void instance(std::shared_ptr<T> service) {
 		static_assert(is_service_single<T>::value, "instance only accept Single Service instance.");
-		
-		call_save_instance(service, tuple_seq<parent_types<T>>());
+
+		call_save_instance(std::move(service), tuple_seq<parent_types<T>>{});
 	}
 	
 	template<typename T>
 	void instance() {
 		static_assert(is_service_single<T>::value, "instance only accept Single Service instance.");
-		
+
 		instance(make_service<T>());
 	}
 	
-	template<typename T, disable_if<is_abstract<T>>..., disable_if<is_base_of_container<T>>...>
+	template<typename T, disable_if<is_abstract<T>>* = nullptr, disable_if<is_base_of_container<T>>* = nullptr>
 	std::shared_ptr<T> service() {
 		return get_service<T>();
 	}
 	
-	template<typename T, enable_if<is_container<T>>...>
+	template<typename T, enable_if<is_container<T>>* = nullptr>
 	std::shared_ptr<T> service() {
 		return shared_from_this();
 	}
 	
-	template<typename T, disable_if<is_container<T>>..., enable_if<is_base_of_container<T>>...>
+	template<typename T, disable_if<is_container<T>>* = nullptr, enable_if<is_base_of_container<T>>* = nullptr>
 	std::shared_ptr<T> service() {
 		return std::dynamic_pointer_cast<T>(shared_from_this());
 	}
 	
-	template<typename T, enable_if<is_abstract<T>>...>
+	template<typename T, enable_if<is_abstract<T>>* = nullptr>
 	std::shared_ptr<T> service() {
-		auto it = _services.find(typeid(T).name());
+		auto it = _services.find(&detail::template type_id<T>);
 		
 		if (it != _services.end()) {
-			auto holder = dynamic_cast<detail::InstanceHolder<T>*>(it->second.get());
+			auto holder = static_cast<detail::InstanceHolder<T>*>(it->second.get());
```

Comment:
`dynamic_cast` is slower.
Since we already know that we have found (see line 149) the InstanceHolder for type T, we can use `static_cast` here.